README.md
camera-display

Display images captured by a C328 serial camera. The code runs on
linux systems and has been tested on Ubuntu and debian running on a
beagleboneblack.


debian/changelog
camera-display (0.0.1-1) unstable; urgency=medium

  * Initial release

 -- Greg Green <ggreen@bit-builder.com>  Fri, 27 Nov 2020 11:13:15 -0800


test_camera.cpp
// enviroment variables recognized:
//
// BAUDRATE=xxxx set the baudrate on the serial port, note the camera only recognizes a few values
// DEBUGSERIAL if set to any value will print the command packets and replies to stderr

#include "c328.h"

#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <unistd.h>

int main(int argc, char* argv[])
{
    if (argc != 2)
    {
        std::cerr << "usage: camera_display picture_filename" << std::endl;
        return -1;
    }
    char* base = argv[1];

    std::string baudrate("9600");
    
    // check environment for baudrate
    char* envval = getenv("BAUDRATE");
    if (envval != nullptr)
        baudrate = envval;

    C328 dev("/dev/ttyUSB0", baudrate);

    // check environment for debug print
    envval = getenv("DEBUGSERIAL");
    if (envval != nullptr) {
        std::cerr << "Turning serial command packet printing on" << std::endl;
        dev.set_debug(true);
    }

    dev.sync();

    if (dev.is_connected())
    {
        std::cerr << "Connected to camera\n";
        bool retval = false;
        for (int i=0; i<5 && !retval; i++)
        {
            retval = dev.initial(TwoBitGrayScale, P160x120, J320x240);
        }
        if (!retval)
        {
            std::cerr << "not initialized" << std::endl;
            return -1;
        }
        std::cerr << "initial setup done\n";
        retval = false;
        for (int i=0; i<5 && !retval; i++)
        {
            retval = dev.set_pkg_size();
        }
        if (!retval)
        {
            std::cerr << "set pkg size failed" << std::endl;
            return -1;
        }
        std::cerr << "set pkg size done\n";
        for (int i=0; i<100; i++)
        {
            auto result = dev.jpeg_snapshot();
            if (result.first == 0)
                continue;

            // write the image file to disk
            char fname[1024];
            snprintf(fname, 1024, "%s%02d.jpg", base, i);
            
            FILE* f = fopen(fname, "w");
            if (f == nullptr)
            {
                std::cerr << "Unable to open file: '" << argv[1] << "' for write" << std::endl;
                return -1;
            }
                // write the file
            size_t n = fwrite(result.second, 1, result.first, f);
            if (n != result.first)
            {
                std::cerr << "Error " << errno << " during write of file: " << strerror(errno) << std::endl;
            }
            fclose(f);
            std::cerr << "Image written to file: '" << fname << "'" << std::endl;
            
            delete [] result.second;

            // sleep a bit between snaps
            usleep(20000);
        }
        return 0;
    }
    std::cerr << "Not connected to camera\n";
    return 1;
}


c328.h
#ifndef C328_H_
#define C328_H_

#include <string>
#include <cstring>
#include <utility>
#include <sys/select.h>

// forward class definitions
class C328CommandPacket;

enum ColorType {
    TwoBitGrayScale = 1,
    FourBitGrayScale = 2,
    EightBitGrayScale = 3,
    TwelveBitColor = 5,
    SixteenBitColor = 6,
    JPEG = 7
};

enum PreviewResolution {
    P80x60 = 1,
    P160x120 = 3
};

enum JPEGResolution {
    J80x64 = 1,
    J160x128 = 3,
    J320x240 = 5,
    J640x480 = 7
};
    
enum SnapshotType {
    Compressed = 0,
    Uncompressed = 1
};

enum PictureType {
    Snapshot = 1,
    Preview = 2,
    JpegPreview = 5
};

class C328
{
public:
    explicit C328(const std::string& serport, const std::string& baudrate);
    ~C328();

    // non-hdwr methods
    void set_debug(bool on) {_debug = on;}
    bool is_connected() const;

    // hdwr commands
    bool initial(enum ColorType ct, enum PreviewResolution pr, enum JPEGResolution jr);
    bool power_off();
    bool reset(bool state_only);
    void sync();
    bool set_pkg_size();
    bool snapshot(enum SnapshotType st);
    void get_picture(enum PictureType pt, uint8_t* pixtype, uint32_t* datasz);
    void get_data_packages(uint32_t datasz, uint8_t* pixbuf);
    void get_image_data(uint32_t datasz, uint8_t* pixbuf);
    
    // composite methods
    std::pair<uint32_t, uint8_t*> jpeg_snapshot();
    std::pair<uint32_t, uint8_t*> uncompressed_snapshot();
    std::pair<uint32_t, uint8_t*> jpeg_preview();
    std::pair<uint32_t, uint8_t*> uncompressed_preview();
    
    // serial port
    bool has_recvd_bytes() const;
    
private:

    bool command_response(const C328CommandPacket& pkt);
    bool read_bytes(uint8_t* buf, ssize_t nbytes);
    void close_port();
    bool write_pkt(const C328CommandPacket& pkt);
    std::pair<C328CommandPacket, bool> read_pkt();
    
    int _serial_port;
    bool _is_sync;
    bool _debug;
};

#endif // C328_H_


c328.cpp
#include "c328.h"
#include "c328_cmd.h"

#include <iostream>
#include <cstring>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <unistd.h>

using namespace std;

C328::C328(const string& serport, const string& baudrate)
    : _serial_port(-1), _is_sync(false), _debug(false)
{
    auto baud = B9600;
    
    // check baudrate
    if (baudrate == "19200")
        baud = B19200;
    else if (baudrate == "38400")
        baud = B38400;
    else if (baudrate == "57600")
        baud = B57600;
    else if (baudrate == "115200")
        baud = B115200;
    cerr << "Baudrate: " << baudrate << endl;
    
    _serial_port = open(serport.c_str(), O_RDWR);
    if (_serial_port < 0)
    {
        cerr << "Error " << errno << " from serial port open: " << strerror(errno) << endl;
        return;
    }
    
    struct termios tty;
    if (tcgetattr(_serial_port, &tty) != 0)
    {
        cerr << "Error " << errno << " from tcgetattr: " << strerror(errno) << endl;
        close(_serial_port);
        _serial_port = -1;
        return;
    }
    // set up raw mode / no echo / binary
    tty.c_cflag |= (CLOCAL | CREAD);
    tty.c_lflag &= ~(ICANON | ECHO | ECHOE |
                     ECHOK | ECHONL | ISIG | IEXTEN |
                     ECHOCTL | ECHOKE);
    tty.c_oflag &= ~(OPOST | ONLCR | OCRNL);
    tty.c_iflag &= ~(INLCR | IGNCR | ICRNL | IGNBRK | IUCLC | PARMRK);

    tty.c_cflag |= CS8;
    tty.c_cflag &= ~(CSTOPB);

    tty.c_iflag &= ~(INPCK | ISTRIP);
    tty.c_cflag &= ~(PARENB | PARODD | CMSPAR);

    tty.c_iflag &= ~(IXON | IXOFF | IXANY);
    tty.c_cflag &= ~(CRTSCTS);
    
    // read 1 char blocking
    tty.c_cc[VMIN] = 1;
    tty.c_cc[VTIME] = 0;

    // baud rate
    cfsetispeed(&tty, baud);
    cfsetospeed(&tty, baud);

    // save tty settings
    if (tcsetattr(_serial_port, TCSANOW, &tty) != 0)
    {
        cerr << "Error " << errno << " from tcsettattr: " << strerror(errno) << endl;
        close(_serial_port);
        _serial_port = -1;
        return;
    }
    
    // finished
    cerr << "Opened serial port '" << serport.c_str() << "' at " << baudrate << " baud" << endl;
}

C328::~C328()
{
    if (_serial_port >= 0)
        close(_serial_port);
}

void C328::close_port()
{
    close(_serial_port);
    _serial_port = -1;
    _is_sync = false;
}

bool C328::write_pkt(const C328CommandPacket& pkt)
{
    ssize_t n = write(_serial_port, pkt.packet(), 6);
    if (n < 0)
    {
        cerr << "Error " << errno << " during serial port write: " << strerror(errno) << endl;
        close_port();
        return false;
    }
    else if (n != 6)
    {
        cerr << "Unable to write packet to port\n";
        close_port();
        return false;
    }
    if (_debug)
        pkt.debug(true);
    return true;
}

bool C328::read_bytes(uint8_t* buf, ssize_t nbytes)
{
    ssize_t count = 0;
    while (count != nbytes)
    {
        ssize_t n = read(_serial_port, buf + count, nbytes - count);
        if (n < 0)
        {
            if (errno == EINTR)
                continue;
            cerr << "Error " << errno << " during serial port read: " << strerror(errno) << endl;
            close_port();
            return false;
        }
        count += n;
    }
    return true;
}

std::pair<C328CommandPacket,bool> C328::read_pkt()
{
    uint8_t buf[6];
    bool ret = read_bytes(buf, 6);
    C328CommandPacket pkt(buf);
    if (_debug)
        pkt.debug(false);
    return make_pair(pkt, ret);
}

bool C328::has_recvd_bytes() const
{
    if (_serial_port < 0)
        return false;

    fd_set readfds;
    FD_ZERO(&readfds);
    // add to select set
    FD_SET(_serial_port, &readfds);
    
    // set timeout to 0 seconds
    struct timeval timeout;
    timeout.tv_sec = 0;
    timeout.tv_usec = 0;

    int retval = select(_serial_port + 1, &readfds, NULL, NULL, &timeout);
    // cerr << "select: " << retval << endl;
    return retval == 1;
}

// implements basic pattern, where a command is sent and an ack is received
bool C328::command_response(const C328CommandPacket& cmdpkt)
{
    if (_serial_port < 0)
        return false;

    uint8_t commandid = cmdpkt.packet()[1];
    
    if (!write_pkt(cmdpkt))
        return false;
    // read the reply
    auto ackval = read_pkt();
    if (!ackval.second)
    {
        cerr << "Camera did not ack" << endl;
        return false;
    }
    else if (ackval.first.is_nak())
    {
        struct NakFields nf;
        ackval.first.nak_fields(&nf);
        cerr << "Camera error command " << static_cast<int>(commandid)
                 << " [" << static_cast<int>(nf.errcode)
             << "]: " << ackval.first.nak_error() << endl;
        return false;
    }
    else if(ackval.first.is_sync())
    {
        cerr << "Camera responded with sync for command "
             << static_cast<int>(commandid) << endl;
    }
    else if (ackval.first.packet()[2] != commandid)
    {
        cerr << "Camera responded with different commandid " << static_cast<int>(ackval.first.packet()[2])
             << " instead of " << static_cast<int>(commandid) << endl;
        //return false;
    }
    return true;
}

bool C328::reset(bool state_only)
{
    C328CommandPacket resetpkt(0x08, state_only ? 0x01 : 0x00, 0x00, 0x00, 0xFF);
    return command_response(resetpkt);
}

bool C328::power_off()
{
    C328CommandPacket pwrpkt(0x09, 0x00, 0x00, 0x00, 0x00);
    return command_response(pwrpkt);
}

void C328::sync()
{
    if (_serial_port < 0)
        return;

    C328CommandPacket sync(0x0D, 0x00, 0x00, 0x00, 0x00);
    for (int i=0; i<200; i++)
    {
        if (!write_pkt(sync))
            return;
        if (has_recvd_bytes())
            break;
        if (usleep(10000) != 0) {
            cerr << "Error " << errno << " during sleep: " << strerror(errno) << endl;
            close_port();
            return;
        }
    }
    if (!has_recvd_bytes())
        return;
    
    // read the reply
    auto ackval = read_pkt();
    if (!ackval.second)
    {
        cerr << "Camera did not send ack packet" << endl;
        return;
    }
    else if (!ackval.first.is_ack() && ackval.first.is_nak())
    {
        struct NakFields nf;
        ackval.first.nak_fields(&nf);
        cerr << "Camera error[" << static_cast<int>(nf.errcode)
             << "]: " << ackval.first.nak_error() << endl;
        return;
    }
    // now read the sync
    auto syncval = read_pkt();
    if (!syncval.second || !syncval.first.is_sync())
    {
        cerr << "Camera did not send sync packet" << endl;
        return;
    }
    // ack the ack
    C328CommandPacket ack(0x0E, 0x0D, 0x00, 0x00, 0x00);
    if (!write_pkt(ack))
        return;
    _is_sync = true;

    // sleep for 2s to let camera stabilize
    if (usleep(2000000) != 0)
    {
        cerr << "Error " << static_cast<int>(errno) << " during sleep: " << strerror(errno) << endl;
        close_port();
        return;
    }
}

bool C328::initial(enum ColorType ct, enum PreviewResolution pr, enum JPEGResolution jr)
{
    C328CommandPacket ini(0x01, 0x00, static_cast<uint8_t>(ct),
                          static_cast<uint8_t>(pr),
                          static_cast<uint8_t>(jr));
    return command_response(ini);
}

// hardcoded for 512 bytes
bool C328::set_pkg_size()
{
    C328CommandPacket setpkgsize(0x06, 0x08, 0x00, 0x02, 0x00);
    return command_response(setpkgsize);
}

bool C328::snapshot(enum SnapshotType st)
{
    C328CommandPacket snap(0x05, static_cast<uint8_t>(st), 0x00, 0x00, 0x00);
    return command_response(snap);
}

void C328::get_picture(enum PictureType pt, uint8_t* pixtype, uint32_t* datasz)
{
    if (_serial_port < 0)
        return;

    *pixtype = 0;
    *datasz = 0;
    
    C328CommandPacket gpix(0x04, static_cast<uint8_t>(pt), 0x00, 0x00, 0x00);
    while (true)
    {
        bool response = command_response(gpix);
        if (response)
            break;
        if(usleep(10000) != 0)
        {
            cerr << "Error " << errno << " during sleep: " << strerror(errno) << endl;
            close_port();
                return;
        }
    }

    // now get data
    auto datval = read_pkt();
    if (!datval.second)
    {
        cerr << "Camera failed to send data packet" << endl;
        return;
    }
    if (datval.first.is_data())
    {
        struct DataFields nf;
        datval.first.data_fields(&nf);
        *datasz = nf.datasz;
        *pixtype = nf.pixtype;
    }
    else if (datval.first.is_nak())
    {
        struct NakFields nf;
        datval.first.nak_fields(&nf);
        cerr << "Camera error[" << static_cast<int>(nf.errcode)
             << "] snapshot: " << datval.first.nak_error() << endl;
        return;
    }
    if(usleep(10000) != 0)
    {
        cerr << "Error " << errno << " during sleep: " << strerror(errno) << endl;
        close_port();
        return;
    }
}

void C328::get_data_packages(uint32_t datasz, uint8_t* pixbuf)
{
    if (_serial_port < 0)
        return;

    int count = 0;
    uint32_t bytecount = 0;
    uint8_t buf[512];
    while (true)
    {
        C328CommandPacket pkgack(0x0E, 0x00, 0x00, (count & 0xFF), (count & 0xFF00) >> 8);
        if (!write_pkt(pkgack))
            return;
        // read header
        if (!read_bytes(buf, 4))
        {
            cerr << "Error while reading data bytes" << endl;
            return;
        }
        uint16_t pkgid = static_cast<uint16_t>((buf[1] << 8) | buf[0]);
        uint16_t pkgsz = static_cast<uint16_t>((buf[3] << 8) | buf[2]);
        //cerr << "Package " << dec << pkgid << " with " << pkgsz << " bytes" << endl;
        if (!read_bytes(buf + 4, pkgsz + 2))
        {
            cerr << "Error while reading data bytes" << endl;
            return;
        }
        uint16_t chksum = static_cast<uint16_t>((buf[pkgsz+5] << 8) | buf[pkgsz+4]);
        //cerr << "Package checksum: " << hex << chksum;
        int sum = 0;
        for (int i=0; i<pkgsz+4; i++)
            sum += buf[i];
        uint16_t calc_chksum = sum & 0xFF;
        //cerr << " calculated: " << hex << calc_chksum << endl;
        if (chksum != calc_chksum)
        {
            cerr << "Error with package " << pkgid << " checksum" << endl;
            return;
        }
        memcpy(pixbuf + bytecount, &buf[4], pkgsz);
        bytecount += pkgsz;
        if (bytecount >= datasz)
            break;
        count++;
        // put in a short sleep after each package
        if(usleep(1000) != 0)
        {
            cerr << "Error " << errno << " during sleep: " << strerror(errno) << endl;
            close_port();
            return;
        }
    }
    // send final ack
    C328CommandPacket ack(0x0E, 0x00, 0x00, 0xF0, 0xF0);
    write_pkt(ack);
}

void C328::get_image_data(uint32_t datasz, uint8_t* pixbuf)
{
    if (_serial_port < 0)
        return;

    if (!read_bytes(pixbuf, datasz))
    {
        cerr << "Error while reading image data" << endl;
        return;
    }
    // send final ack
    C328CommandPacket ack(0x0E, 0x0A, 0x00, 0x00, 0x00);
    write_pkt(ack);
}

bool C328::is_connected() const
{
    return _serial_port >= 0 && _is_sync;
}

std::pair<uint32_t, uint8_t*> C328::jpeg_snapshot()
{
    if (!snapshot(Compressed))
        if (!snapshot(Compressed))
            return make_pair(0, nullptr);

    uint8_t pixtype;
    uint32_t datasz;

    get_picture(Snapshot, &pixtype, &datasz);
    if (pixtype == 0)
    {
        std::cerr << "Error during get picture\n";
        return make_pair(0, nullptr);
    }
    //std::cerr << "Pixtype: " << static_cast<int>(pixtype) << " size: " << datasz << std::endl;

    // allocate a buffer to hold the pix
    uint8_t* pixbuf = new uint8_t[datasz];

    get_data_packages(datasz, pixbuf);
    
    return make_pair(datasz, pixbuf);
}

std::pair<uint32_t, uint8_t*> C328::uncompressed_snapshot()
{
    if (!snapshot(Uncompressed))
        return make_pair(0, nullptr);

    uint8_t pixtype;
    uint32_t datasz;

    get_picture(Snapshot, &pixtype, &datasz);
    if (pixtype == 0)
    {
        std::cerr << "Error during get picture\n";
        return make_pair(0, nullptr);
    }
    //std::cerr << "Pixtype: " << static_cast<int>(pixtype) << " size: " << datasz << std::endl;

    // allocate a buffer to hold the pix
    uint8_t* pixbuf = new uint8_t[datasz];

    get_image_data(datasz, pixbuf);
    
    return make_pair(datasz, pixbuf);
}

std::pair<uint32_t, uint8_t*> C328::jpeg_preview()
{
    uint8_t pixtype;
    uint32_t datasz;

    get_picture(JpegPreview, &pixtype, &datasz);
    if (pixtype == 0)
    {
        std::cerr << "Error during get picture\n";
        return make_pair(0, nullptr);
    }
    //std::cerr << "Pixtype: " << static_cast<int>(pixtype) << " size: " << datasz << std::endl;

    // allocate a buffer to hold the pix
    uint8_t* pixbuf = new uint8_t[datasz];

    get_data_packages(datasz, pixbuf);
    
    return make_pair(0, nullptr);
}

std::pair<uint32_t, uint8_t*> C328::uncompressed_preview()
{
    uint8_t pixtype;
    uint32_t datasz;

    get_picture(Preview, &pixtype, &datasz);
    if (pixtype == 0)
    {
        std::cerr << "Error during get picture\n";
        return make_pair(0, nullptr);
    }
    //std::cerr << "Pixtype: " << static_cast<int>(pixtype) << " size: " << datasz << std::endl;

    // allocate a buffer to hold the pix
    uint8_t* pixbuf = new uint8_t[datasz];

    get_image_data(datasz, pixbuf);
    
    return make_pair(datasz, pixbuf);
}


c328_cmd.h
#ifndef C328_CMD_H_
#define C328_CMD_H_

#include <cstdint>
#include <string>

struct AckFields
{
    uint8_t cmdid;
    uint8_t counter;
    uint16_t pkgid;
};

struct NakFields
{
    uint8_t counter;
    uint8_t errcode;
};

struct DataFields
{
    uint8_t pixtype;
    uint32_t datasz;
};

class C328CommandPacket
{
public:
    C328CommandPacket() = default;
    explicit C328CommandPacket(uint8_t b2, uint8_t b3, uint8_t b4, uint8_t b5, uint8_t b6);
    explicit C328CommandPacket(const uint8_t* buf);
    ~C328CommandPacket() {;}

    const uint8_t* packet() const {return _data;}

    bool is_sync() const;
    bool is_nak() const;
    bool is_ack() const;
    bool is_data() const;
    
    void ack_fields(struct AckFields* fld) const;
    void nak_fields(struct NakFields* fld) const;
    void data_fields(struct DataFields* fld) const;

    // dump to std::err, dir is True for write
    void debug(bool dir) const;

    const std::string& nak_error() const;
    
private:
    uint8_t _data[6];
};

#endif // C328_CMD_H_


c328_cmd.cpp
#include "c328_cmd.h"
#include <cstring>
#include <cstdio>

using namespace std;

C328CommandPacket::C328CommandPacket(uint8_t b2, uint8_t b3, uint8_t b4, uint8_t b5, uint8_t b6)
{
    _data[0] = 0xAA;
    _data[1] = b2;
    _data[2] = b3;
    _data[3] = b4;
    _data[4] = b5;
    _data[5] = b6;
}

C328CommandPacket::C328CommandPacket(const uint8_t* buf)
{
    memcpy(_data, buf, 6);
}

bool C328CommandPacket::is_sync() const
{
    return _data[0] == 0xAA && _data[1] == 0x0D;
}

bool C328CommandPacket::is_ack() const
{
    return _data[0] == 0xAA && _data[1] == 0x0E;
}

bool C328CommandPacket::is_nak() const
{
    return _data[0] == 0xAA && _data[1] == 0x0F;
}

bool C328CommandPacket::is_data() const
{
    return _data[0] == 0xAA && _data[1] == 0x0A;
}

void C328CommandPacket::ack_fields(struct AckFields* fld) const
{
    fld->cmdid = _data[2];
    fld->counter = _data[3];
    fld->pkgid = (_data[5] << 8) | _data[4];
}

void C328CommandPacket::nak_fields(struct NakFields* fld) const
{
    fld->counter = _data[3];
    fld->errcode = _data[4];
}

void C328CommandPacket::data_fields(struct DataFields* fld) const
{
    fld->pixtype = _data[2];
    fld->datasz = (_data[5] << 16) | (_data[4] << 8) | _data[3];
}

void C328CommandPacket::debug(bool dir) const
{
    fprintf(stderr, "%s %02x %02x %02x %02x %02x %02x\n", dir ? "->" : "<-",
            _data[0], _data[1], _data[2], _data[3], _data[4], _data[5]);
}

struct err_to_msg 
{
    uint8_t errcode;
    std::string errmsg;
};

const struct err_to_msg nak_errs[] = {
    {0x00, "Not a known NAK errorcode, or not a NAK reply, no error message available"},
    {0x01, "Picture Type Error"},
    {0x02, "Picture Up Scale"},
    {0x03, "Picture Scale Error"},
    {0x04, "Unexpected Reply"},
    {0x05, "Send Picture Timeout"},
    {0x06, "Unexpected Command"},
    {0x07, "SRAM JPEG Type Error"},
    {0x08, "SRAM JPEG Size Error"},
    {0x09, "Picture Format Error"},
    {0x0A, "Picture Size Error"},
    {0x0B, "Parameter Error"},
    {0x0C, "Send Register Timeout"},
    {0x0D, "Command ID Error"},
    {0x0F, "Picture Not Ready"},
    {0x10, "Transfer Package Number Error"},
    {0x11, "Set Transfer Package Size Wrong"},
    {0xF0, "Command Header Error"},
    {0xF1, "Command Length Error"},
    {0xF5, "Send Picture Error"},
    {0xFF, "Send Command Error"},
};

const std::string& C328CommandPacket::nak_error() const
{
    if (is_nak()) {
        for (size_t i=0; i<sizeof(nak_errs); i++)
            if (nak_errs[i].errcode == _data[4])
                return nak_errs[i].errmsg;
    }
    return nak_errs[0].errmsg;
}


main.cpp
// enviroment variables recognized:
//
// BAUDRATE=xxxx set the baudrate on the serial port, note the camera only recognizes a few values
// DEBUGSERIAL if set to any value will print the command packets and replies to stderr

#include "c328.h"

#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include <cstring>

// For compilers that support precompilation, includes "wx/wx.h".
#include "wx/wxprec.h"


#ifndef WX_PRECOMP
    #include "wx/wx.h"
#endif

#include "wx/image.h"
#include "wx/mstream.h"

using namespace std;

class MyApp : public wxApp
{
public:
    virtual bool OnInit();
};

class MyFrame : public wxFrame
{
public:
    MyFrame();

    void OnPaint(wxPaintEvent& event);
    void OnTimer(wxTimerEvent& event);

private:
    
    void get_image();
    void load_image(uint8_t* data, size_t datasz);
    void OnHello(wxCommandEvent& event);
    void OnAbout(wxCommandEvent& event);

    C328* _dev;
    wxBitmap _cmyk_jpeg;
    wxTimer _timer;
    
    wxDECLARE_DYNAMIC_CLASS(MyFrame);
    wxDECLARE_EVENT_TABLE();
};

enum
{
    ID_Hello = 1,
    ID_Timer = 2
};

// main function
wxIMPLEMENT_APP(MyApp);

bool MyApp::OnInit()
{
    wxInitAllImageHandlers();

    MyFrame *frame = new MyFrame();
    frame->Show(true);

    return true;
}

MyFrame::MyFrame()
    : wxFrame(nullptr, wxID_ANY, "camera_display"), _dev(nullptr), _timer(this, ID_Timer)
{
    wxMenu *menuFile = new wxMenu;
    menuFile->Append(ID_Hello, "&Hello...\tCtrl-H",
                     "Help string shown in status bar for this menu item");
    menuFile->AppendSeparator();
    menuFile->Append(wxID_EXIT);
    wxMenu *menuHelp = new wxMenu;
    menuHelp->Append(wxID_ABOUT);
    wxMenuBar *menuBar = new wxMenuBar;
    menuBar->Append(menuFile, "&File");
    menuBar->Append(menuHelp, "&Help");
    SetMenuBar(menuBar);
    auto statusBar = CreateStatusBar();
    SetStatusText("Welcome to wxWidgets!");

    Bind(wxEVT_MENU, &MyFrame::OnHello, this, ID_Hello);
    Bind(wxEVT_MENU, &MyFrame::OnAbout, this, wxID_ABOUT);
    Bind(wxEVT_MENU, [=](wxCommandEvent&) { Close(true); }, wxID_EXIT);

    std::string baudrate("9600");
    
    // check environment for baudrate
    char* envval = getenv("BAUDRATE");
    if (envval != nullptr)
        baudrate = envval;

    _dev = new C328("/dev/ttyUSB0", baudrate);

    // check environment for debug print
    envval = getenv("DEBUGSERIAL");
    if (envval != nullptr) {
        SetStatusText("Turning serial command packet printing on");
        _dev->set_debug(true);
    }

    _dev->sync();
    _timer.Start(10000);
    //get_image();

    SetClientSize(320,200+menuBar->m_height+statusBar->m_height);

    Show();
}

void MyFrame::get_image()
{
    //load_image(nullptr, 0);
    //return;
    if (_dev->is_connected())
    {
        std::cerr << "Connected to camera\n";
        bool retval = false;
        for (int i=0; i<5 && !retval; i++)
        {
            retval = _dev->initial(TwoBitGrayScale, P160x120, J320x240);
        }
        if (!retval)
        {
            _dev->reset(true);
            std::cerr << "not initialized" << std::endl;
            return;
        }
        std::cerr << "initial setup done\n";
        retval = false;
        for (int i=0; i<5 && !retval; i++)
        {
            retval = _dev->set_pkg_size();
        }
        if (!retval)
        {
            _dev->reset(true);
            std::cerr << "set pkg size failed" << std::endl;
            return;
        }
        std::cerr << "set pkg size done\n";
        auto result = _dev->jpeg_snapshot();
        if (result.first != 0)
        {
            load_image(result.second, result.first);
            delete [] result.second;
        }
    }
    else
        _dev->reset(true);
    Fit();
}

void MyFrame::load_image(uint8_t* data, size_t datasz)
{
    wxMemoryInputStream idstream(data, datasz);
    
    wxImage image;
    if (!image.LoadFile(idstream, wxBITMAP_TYPE_JPEG)) {
//    if (!image.LoadFile("/spare/export/src/boat-os/build-camera-display/pix00.jpg")) {
        cerr << "Can't load image from data" << endl;
        return;
    }
    _cmyk_jpeg = wxBitmap(image);
}

void MyFrame::OnTimer(wxTimerEvent& WXUNUSED(event))
{
    get_image();
}

void MyFrame::OnPaint(wxPaintEvent& WXUNUSED(event))
{
    wxPaintDC dc(this);
    PrepareDC(dc);
    if (_cmyk_jpeg.IsOk())
        dc.DrawBitmap(_cmyk_jpeg, 0, 0);
}
    
void MyFrame::OnAbout(wxCommandEvent& event)
{
    wxMessageBox("This is a wxWidgets Hello World example",
                 "About Hello World", wxOK | wxICON_INFORMATION);
}

void MyFrame::OnHello(wxCommandEvent& event)
{
    wxLogMessage("Hello world from wxWidgets!");
}

IMPLEMENT_DYNAMIC_CLASS(MyFrame, wxFrame)
wxBEGIN_EVENT_TABLE(MyFrame, wxFrame)
    EVT_PAINT(MyFrame::OnPaint)
    EVT_TIMER(ID_Timer, MyFrame::OnTimer)
wxEND_EVENT_TABLE()
